### ISSUE TITLE
Weird behaviour when using transform on new data

### ISSUE THREAD

**Issue by generalizedeigenvector** [lmcinnes/umap#358]
I have encountered a very strange behaviour when trying to use UMAP for a fairly simple feature learning scenario.

As a minimal example, I took the digits code from the tutorial and modified it a little in the following way instead of the usual split:

- The training points are now those whose labels are < 9.
- The test points are those with label 9.

This is the code:
`import numpy as np
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC

import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import umap

sns.set(context='notebook', style='white', rc={'figure.figsize':(14,10)})

digits = load_digits()

X_train, X_test, y_train, y_test = train_test_split(digits.data,
                                                    digits.target,
                                                    stratify=digits.target,
                                                    random_state=42)
print(X_train.shape, X_test.shape)

i1 = np.where(digits.target<9)[0]
i2 = np.where(digits.target==9)[0]

X_train = digits.data[i1,:]
y_train = digits.target[i1]
X_test = digits.data[i2,:]
y_test = digits.target[i2]

print(X_train.shape, X_test.shape)

trans = umap.UMAP(n_neighbors=15, random_state=42).fit(X_train, y_train)

plt.scatter(trans.embedding_[:, 0], trans.embedding_[:, 1], s= 5, c=y_train, cmap='Spectral')
plt.title('Embedding of the training set by UMAP', fontsize=24);

test_embedding = trans.transform(X_test)

alldata = np.concatenate((trans.embedding_, test_embedding))
yy = np.concatenate((y_train, y_test))

plt.scatter(alldata[:, 0], alldata[:, 1], s= 5, c=yy, cmap='Spectral')
plt.title('Now apply the transform to the test and plot all together:', fontsize=24);`

The result is very strange:
![image](https://user-images.githubusercontent.com/59834480/73955744-818a9680-490c-11ea-9efb-60b4db8db291.png)

It seems as if the training data is somehow mixed up with the transformation of the completely unrelated test data. The same thing happens with other data as well.

Is this a bug?

**Comment by lmcinnes** [comment]
Sorry that I haven't had time to look at this -- I'll try to get to it soon.

**Comment by generalizedeigenvector** [comment]
Thanks, looking forward to it!

**Comment by lmcinnes** [comment]
So I think the issue here is that the transform can't learn "new" unseen structures, it can only relate new data to the training data it has seen. In this case that means that since the test data is all the 9 digits which are quite different from the training data, it is struggling to do much with them -- the training data just isn't a decent representative sample of the data that is showing up later in the test set to be transformed in. Since the 9s look somewhat like 3 or 5 they seem to mostly get folded in there, but scattered with other clusters as well.

In summary this isn't a bug, so much as a fundamental limitation of the technique. I would expect, however, that you might see similar results if you tried PCA or autoencoders with a similar test/train split.

**Comment by sleighsoft** [comment]
@generalizedeigenvector and @lmcinnes I think this comparison (umap, pca, autoencoder) and a short discussion would be cool to have in the https://github.com/lmcinnes/umap/tree/master/notebooks section. Maybe you could create one @generalizedeigenvector
